presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/HashGenerationOptimizer.java

```diff
@@ -314,6 +318,11 @@ public PlanWithProperties visitTopNRowNumber(TopNRowNumberNode node, HashComputa
                     child.getHashVariables());
         }
 
+        private boolean skipHashComputeForJoinInput(PlanNode node, Optional<HashComputation> hashComputation, HashComputationSet parentPreference)
+        {
+            return node instanceof TableScanNode && hashComputation.isPresent() && hashComputation.get().isSingleBigIntVariable() && !parentPreference.getHashes().contains(hashComputation.get());
```

Comment:
I'm curious why we would ever add this hash computation when the parent does not require it. I.e. wouldn't this check simply always be 
"return hashComputation.isPresent() && !parentPreference.getHashes().contains(hashComputation.get());"

Reply:
This optimization is based on our observation, where TableScan below join is significantly than ScanProject (here project is for hash generation) for big int join key. Do not observe the same for other cases.